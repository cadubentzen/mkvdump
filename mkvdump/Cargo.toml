[package]
name = "mkvdump"
version = "0.5.0"
authors = ["Carlos Bentzen <cadubentzen@gmail.com>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "MKV and WebM parser CLI tool"
repository = "https://github.com/cadubentzen/mkvdump"
keywords = ["mkv", "matroska", "webm", "cli", "ebml"]
categories = ["command-line-utilities", "multimedia", "parser-implementations"]
exclude = ["tests/*", ".github/*"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
mkvparser = { path = "../mkvparser", version = "0.4.4" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
serde_yaml = "0.9"
clap = { version = "4", features = ["derive"] }

[dev-dependencies]
assert_cmd = "2"
insta = { version = "1.18.1", features = ["yaml"] }
