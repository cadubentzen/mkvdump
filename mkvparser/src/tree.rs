///! The tree module contains helpers for building tree
///! structures from parsed elements
use serde::Serialize;

use crate::{Body, Element, Header, Id};

/// A Master Element that owns its children for diplaying
/// it in an element tree
#[derive(Debug, PartialEq, Serialize)]
pub struct MasterElement {
    #[serde(flatten)]
    header: Header,
    children: Vec<ElementTree>,
}

/// An Element Tree can either be a leaf or a Master
/// element.
#[derive(Debug, PartialEq, Serialize)]
#[serde(untagged)]
pub enum ElementTree {
    /// A Normal Element that represents a leaf in the tree
    Normal(Element),
    /// A Master Element contains more elements.
    Master(MasterElement),
}

impl Id {
    fn can_be_children_of(&self, other: &Id) -> bool {
        !matches!((self, other), (Id::Cluster, Id::Cluster) | (Id::Ebml, _))
    }
}

/// Build element trees from a series of elements
pub fn build_element_trees(elements: &[Element]) -> Vec<ElementTree> {
    let mut trees = Vec::<ElementTree>::new();

    let mut index = 0;
    while index < elements.len() {
        let element = &elements[index];
        match element.body {
            Body::Master => {
                // parse_header() already handles Unknown sizes.
                let mut size_remaining = element.header.body_size.unwrap_or(usize::MAX);

                let mut children = Vec::<Element>::new();
                while size_remaining > 0 {
                    index += 1;

                    if let Some(next_child) = elements.get(index) {
                        if !next_child.header.id.can_be_children_of(&element.header.id) {
                            index -= 1;
                            break;
                        }

                        size_remaining -= if let Body::Master = next_child.body {
                            // Master elements' body size should not count in the recursion
                            // as the children would duplicate the size count, so
                            // we only consider the header size on the calculation.
                            next_child.header.header_size
                        } else {
                            next_child
                                .header
                                .size
                                .expect("Only Master elements can have unknown size")
                        };
                        children.push(next_child.clone());
                    } else {
                        // Elements have ended before reaching the size of the master element
                        break;
                    }
                }
                trees.push(ElementTree::Master(MasterElement {
                    header: element.header.clone(),
                    children: build_element_trees(&children),
                }));
            }
            _ => {
                trees.push(ElementTree::Normal(element.clone()));
            }
        }
        index += 1;
    }
    trees
}

#[cfg(test)]
mod tests {
    use crate::Unsigned;

    use super::*;

    #[test]
    fn test_build_element_trees() {
        let elements = [
            Element {
                header: Header::new(Id::Ebml, 5, 31),
                body: Body::Master,
            },
            Element {
                header: Header::new(Id::EbmlVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(1)),
            },
            Element {
                header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(8)),
            },
            Element {
                header: Header::new(Id::DocType, 3, 4),
                body: Body::String("webm".to_string()),
            },
            Element {
                header: Header::new(Id::DocTypeVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(4)),
            },
            Element {
                header: Header::new(Id::DocTypeReadVersion, 3, 1),
                body: Body::Unsigned(Unsigned::Standard(2)),
            },
        ];

        let expected = vec![ElementTree::Master(MasterElement {
            header: Header::new(Id::Ebml, 5, 31),
            children: vec![
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(1)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxIdLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::EbmlMaxSizeLength, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(8)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocType, 3, 4),
                    body: Body::String("webm".to_string()),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(4)),
                }),
                ElementTree::Normal(Element {
                    header: Header::new(Id::DocTypeReadVersion, 3, 1),
                    body: Body::Unsigned(Unsigned::Standard(2)),
                }),
            ],
        })];

        assert_eq!(build_element_trees(&elements), expected);
    }
}
