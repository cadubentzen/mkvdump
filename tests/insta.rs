use mkvdump::parse_elements_from_file;
use mkvparser::tree::build_element_trees;

const BUFFER_SIZE: u64 = 8192;

macro_rules! snapshot_test {
    ($test_name:ident, $filename:expr) => {
        #[test]
        fn $test_name() -> anyhow::Result<()> {
            let elements =
                parse_elements_from_file(concat!("tests/inputs/", $filename), false, BUFFER_SIZE)?;
            insta::assert_yaml_snapshot!(build_element_trees(&elements));
            Ok(())
        }
    };
}

snapshot_test!(test_parse_incomplete_file_should_not_hang, "incomplete.hdr");
snapshot_test!(test_parse_header_encrypted, "encrypted.hdr");

// File was generated with:
// ffmpeg -f lavfi -i testsrc -c:v libx264 -frames:v 2 -metadata creation_time="2022-08-11T08:27:15Z" -f matroska test.mkv
snapshot_test!(test_parse_file_with_dateutc, "dateutc.mkv");

// Tests from Matroska test suite
snapshot_test!(test1, "matroska-test-suite/test1.mkv");
snapshot_test!(test2, "matroska-test-suite/test2.mkv");
snapshot_test!(test3, "matroska-test-suite/test3.mkv");
snapshot_test!(test4, "matroska-test-suite/test4.mkv");
snapshot_test!(test5, "matroska-test-suite/test5.mkv");
snapshot_test!(test6, "matroska-test-suite/test6.mkv");
snapshot_test!(test7, "matroska-test-suite/test7.mkv");
snapshot_test!(test8, "matroska-test-suite/test8.mkv");

snapshot_test!(
    test_two_inits_segment_unknown_size,
    "two_inits_segment_unknown_size.webm"
);
snapshot_test!(
    test_init_after_cluster_unknown_size,
    "init_after_cluster_unknown_size.webm"
);
