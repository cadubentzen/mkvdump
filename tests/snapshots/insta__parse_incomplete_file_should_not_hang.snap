---
source: tests/insta.rs
expression: build_element_trees(&elements)
---
- id: EBML
  header_size: 5
  size: 40
  children:
    - id: EBMLVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLReadVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLMaxIDLength
      header_size: 3
      size: 4
      value: 4
    - id: EBMLMaxSizeLength
      header_size: 3
      size: 4
      value: 8
    - id: DocType
      header_size: 3
      size: 11
      value: matroska
    - id: DocTypeVersion
      header_size: 3
      size: 4
      value: 4
    - id: DocTypeReadVersion
      header_size: 3
      size: 4
      value: 2
- id: Segment
  header_size: 12
  size: 6618
  children:
    - id: SeekHead
      header_size: 5
      size: 70
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[c7 c3 2b 17]"
        - id: Seek
          header_size: 3
          size: 14
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Info
            - id: SeekPosition
              header_size: 3
              size: 4
              value: 229
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Tracks
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 272
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Tags
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 343
        - id: Seek
          header_size: 3
          size: 15
          children:
            - id: SeekID
              header_size: 3
              size: 7
              value: Cues
            - id: SeekPosition
              header_size: 3
              size: 5
              value: 6560
    - id: Void
      header_size: 9
      size: 159
      value: ~
    - id: Info
      header_size: 5
      size: 43
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[8e 09 74 4b]"
        - id: TimestampScale
          header_size: 4
          size: 7
          value: 1000000
        - id: MuxingApp
          header_size: 3
          size: 7
          value: Lavf
        - id: WritingApp
          header_size: 3
          size: 7
          value: Lavf
        - id: Duration
          header_size: 3
          size: 11
          value: 10000
    - id: Tracks
      header_size: 5
      size: 71
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[9f 90 75 5f]"
        - id: TrackEntry
          header_size: 9
          size: 60
          children:
            - id: TrackNumber
              header_size: 2
              size: 3
              value: 1
            - id: TrackUID
              header_size: 3
              size: 4
              value: 1
            - id: FlagLacing
              header_size: 2
              size: 3
              value: 0
            - id: Language
              header_size: 4
              size: 7
              value: und
            - id: CodecID
              header_size: 2
              size: 7
              value: V_VP9
            - id: TrackType
              header_size: 2
              size: 3
              value: video
            - id: DefaultDuration
              header_size: 4
              size: 8
              value: 40000000
            - id: Video
              header_size: 9
              size: 16
              children:
                - id: PixelWidth
                  header_size: 2
                  size: 4
                  value: 360
                - id: PixelHeight
                  header_size: 2
                  size: 3
                  value: 240
    - id: Tags
      header_size: 6
      size: 142
      children:
        - id: CRC-32
          header_size: 2
          size: 6
          value: "[19 f8 55 eb]"
        - id: Tag
          header_size: 10
          size: 62
          children:
            - id: Targets
              header_size: 10
              size: 14
              children:
                - id: TagTrackUID
                  header_size: 3
                  size: 4
                  value: 1
            - id: SimpleTag
              header_size: 10
              size: 38
              children:
                - id: TagName
                  header_size: 3
                  size: 10
                  value: ENCODER
                - id: TagString
                  header_size: 3
                  size: 18
                  value: Lavc libvpx-vp9
        - id: Tag
          header_size: 10
          size: 68
          children:
            - id: Targets
              header_size: 10
              size: 14
              children:
                - id: TagTrackUID
                  header_size: 3
                  size: 4
                  value: 1
            - id: SimpleTag
              header_size: 10
              size: 44
              children:
                - id: TagName
                  header_size: 3
                  size: 11
                  value: DURATION
                - id: Corrupted
                  header_size: 0
                  size: 16
                  value: ~

